cli/lib/kontena/stacks/change_resolver.rb

```diff
@@ -56,41 +56,63 @@ def remaining_stacks
       @remaining_stacks ||= added_stacks + upgraded_stacks
     end
 
+    # @return [Boolean]
+    def safe?
+      removed_stacks.empty? && replaced_stacks.empty? && removed_services.empty?
+    end
+
     def analyze
-      old_names = old_data.keys
-      new_names = new_data.keys
+      old_names = old_data.stack_names
+      new_names = new_data.stack_names
 
       removed_stacks.concat(old_names - new_names)
       added_stacks.concat(new_names - old_names)
-      upgraded_stacks.concat(new_names & old_names)
+      (new_names & old_names).each do |candidate|
+        upgraded_stacks << candidate if stack_upgraded?(candidate)
```

Comment:
Could this be written as something like `upgraded_stacks = (new_names & old_names).select{|x| stack_upgraded?(x)`

Reply:
Yes, makes sense.